lib/inventory/metadata/metadata.go

```diff
@@ -0,0 +1,322 @@
+			return client.Do(req)
+		}
+	}
+	if c.kubeClient == nil {
+		c.kubeClient = getKubeClient()
+	}
+}
+
+// getKubeClient returns a kubernetes client in case the instance is running on
+// kubernetes. It returns nil otherwise.
+func getKubeClient() kubernetes.Interface {
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		log.Debugf("Failed to get kubernetes cluster config: %s", err)
+		return nil
+	}
+	client, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		log.Debugf("Failed to create kubernetes client: %s", err)
+		return nil
+	}
+	return client
+}
+
+// fetch fetches all metadata.
+func (c *fetchConfig) fetch() *Metadata {
+	return &Metadata{
+		OS:                    c.fetchOS(),
+		OSVersion:             c.fetchOSVersion(),
+		HostArchitecture:      c.fetchHostArchitecture(),
+		GlibcVersion:          c.fetchGlibcVersion(),
+		InstallMethods:        c.fetchInstallMethods(),
+		ContainerRuntime:      c.fetchContainerRuntime(),
+		ContainerOrchestrator: c.fetchContainerOrchestrator(),
+		CloudEnvironment:      c.fetchCloudEnvironment(),
+	}
+}
+
+// fetchOS returns the value of GOOS.
+func (c *fetchConfig) fetchOS() string {
+	return runtime.GOOS
+}
+
+// fetchHostArchitecture returns the output of arch.
+func (c *fetchConfig) fetchHostArchitecture() string {
+	arch, err := c.exec("arch")
+	if err != nil {
+		return ""
+	}
+
+	return arch
+}
+
+// fetchInstallMethods returns the list of methods used to install the instance.
+func (c *fetchConfig) fetchInstallMethods() []string {
+	installMethods := []string{}
+	if c.dockerfileInstallMethod() {
+		installMethods = append(installMethods, "dockerfile")
+	}
+	if c.helmKubeAgentInstallMethod() {
+		installMethods = append(installMethods, "helm_kube_agent")
+	}
+	if c.nodeScriptInstallMethod() {
+		installMethods = append(installMethods, "node_script")
+	}
+	if c.systemctlInstallMethod() {
+		installMethods = append(installMethods, "systemctl")
+	}
+	return installMethods
+}
+
+// dockerfileInstallMethod returns true if the instance was installed using our
+// Dockerfile.
+func (c *fetchConfig) dockerfileInstallMethod() bool {
+	return c.getenv("TELEPORT_INSTALL_METHOD_DOCKERFILE") == "true"
+}
+
+// helmKubeAgentInstallMethod returns true if the instance was installed using our
+// Helm chart.
+func (c *fetchConfig) helmKubeAgentInstallMethod() bool {
+	return c.getenv("TELEPORT_INSTALL_METHOD_HELM_KUBE_AGENT") == "true"
+}
+
+// nodeScriptInstallMethod returns true if the instance was installed using our
+// install-node.sh script.
+func (c *fetchConfig) nodeScriptInstallMethod() bool {
+	return c.getenv("TELEPORT_INSTALL_METHOD_NODE_SCRIPT") == "true"
+}
+
+// systemctlInstallMethod returns true if the instance is running using systemctl.
+func (c *fetchConfig) systemctlInstallMethod() bool {
+	out, err := c.exec("systemctl", "status", "teleport.service")
+	if err != nil {
+		return false
+	}
+
+	return strings.Contains(out, "active (running)")
```

Comment:
We should probably replace this with something that uses the systemd dbus API in the future.

Reply:
Thanks! Added this as a possible improvement in https://github.com/gravitational/cloud/issues/3550#issue-1591930330.